### DISCUSSION TITLE
Android build

### DISCUSSION THREAD

**Discussion by AlothNaveen** [open-quantum-safe/liboqs#1833] [Q&A] [answered]
How to compile or build liboqs with android architecture.
$:~/naveen/liboqs/build$ ./build-android.sh /home/bcs/naveen/26.1.10909125 -a armeabi-v7a -b /home/bcs/naveen/liboqs/build/armeabi-v7a -s 21
Valid directory for NDK at /home/bcs/naveen/26.1.10909125
Compiling for ABI armeabi-v7a
Compiling for SDK 21
Cleaning up previous build
Building in directory /home/bcs/naveen/liboqs/build/armeabi-v7a
CMake Error: The source directory "/home/bcs/naveen/liboqs/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
$:~/naveen/liboqs/build$  cmake -GNinja ..
-- The C compiler identification is GNU 9.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test CC_SUPPORTS_WA_NOEXECSTACK
-- Performing Test CC_SUPPORTS_WA_NOEXECSTACK - Success
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Alg enablement unchanged
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found suitable version "1.1.1f", minimum required is "1.1.1")  
-- Looking for aligned_alloc
-- Looking for aligned_alloc - found
-- Looking for posix_memalign
-- Looking for posix_memalign - found
-- Looking for memalign
-- Looking for memalign - found
-- Looking for explicit_bzero
-- Looking for explicit_bzero - found
-- Looking for memset_s
-- Looking for memset_s - not found
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.17") found components: doxygen dot 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/bcs/naveen/liboqs/build
$:~/naveen/liboqs/build$ ./build-android.sh /home/bcs/naveen/26.1.10909125 -a armeabi-v7a -b /home/bcs/naveen/liboqs/build/armeabi-v7a -s 21
Valid directory for NDK at /home/bcs/naveen/26.1.10909125
Compiling for ABI armeabi-v7a
Compiling for SDK 21
Cleaning up previous build
Building in directory /home/bcs/naveen/liboqs/build/armeabi-v7a
-- Alg enablement unchanged
-- Looking for getentropy
-- Looking for getentropy - found
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    ANDROID_ABI
    ANDROID_NATIVE_API_LEVEL
    CMAKE_TOOLCHAIN_FILE

-- Build files have been written to: /home/bcs/naveen/liboqs/build
Error: could not load cache
$:~/naveen/liboqs/build$ ninja
[1171/1171] Linking C executable tests/speed_sig
$:~/naveen/liboqs/build$ ./build-android.sh /home/bcs/naveen/26.1.10909125 -a armeabi-v7a -b /home/bcs/naveen/liboqs/build/armeabi-v7a -s 21
Valid directory for NDK at /home/bcs/naveen/26.1.10909125
Compiling for ABI armeabi-v7a
Compiling for SDK 21
Cleaning up previous build
Building in directory /home/bcs/naveen/liboqs/build/armeabi-v7a
-- Alg enablement unchanged
-- Configuring done
-- Generating done
-- Build files have been written to: /home/bcs/naveen/liboqs/build
Error: could not load cache
$:~/naveen/liboqs/build$

**Comment by SWilson4** [comment]
Hi @AlothNaveen, I'm having trouble parsing the logs you've included above. Could you please flesh out the explanation to help me understand (1) what you're trying to achieve, (2) what commands you're running, and (3) how the behaviour differs from what you're expecting?

**Comment by AlothNaveen** [comment]
can we get liboqs.a file insteas of  .so file with build-android.sh

**Reply by dstebila** [reply]
We don't have much expertise with Android on the project team. You can try modifying [build-android.sh around line 109](https://github.com/open-quantum-safe/liboqs/blob/main/scripts/build-android.sh#L109) to see if that gives you what you need, and if so we'd be happy to receive a pull request.

**Comment by AlothNaveen** [comment]
I tried... It didn't worked for me

**Reply by dstebila** [reply]
@res0nance was one of the most recent people with Android expertise to look at our build system. Perhaps they have some thoughts.

**Comment by baentsch** [comment]
Also tagging @andybrucenet : There's a long-standing PR around (also Android) static builds https://github.com/open-quantum-safe/liboqs/pull/1717; maybe that'd be informative @AlothNaveen . But confirming @dstebila's statement: We have insufficient Android expertise, so any PR and continued contribution for that platform would be _very_ welcome.

**Comment by res0nance** [comment, marked as answer]
I believe to get an `.a` file you only need
```
git clone https://github.com/open-quantum-safe/liboqs
cd liboqs
mkdir build
cd build
cmake -DANDROID_ABI=<ANDROID_ABI> -DCMAKE_TOOLCHAIN_FILE=<PATH TO NDK>/build/cmake/android.toolchain.cmake <ADDITIONAL OQS FLAGS> ..
make all
```
This will produce a `.a` file in the `build/lib` folder.

The reason the build-android.sh file produces an `.so` file is because the `-DBUILD_SHARED_LIBS=ON` flag is included. Turn that flag off and it should produce an `.a` file

**Comment by DeekshantDewari** [comment]
@res0nance hey buddy can you tell me how can i access oqs provider in android . I have compiled openssl oqs provider and liboqs for android but while accessing oqs provider I'm not able to access it , I have made changes in openssl.cnf file , I tired to load oqs manually , I even tried to set environment variable but oqs provider is not accessible .
If you have any solutiion please do tell.
Thanks
